File: docs/palindromes-in-peg/scaffold_circuit_structs.py
"""Persistent stack and tape operations emitted as finite scaffold equations.

All loops here range over a fixed construction-time cell layout. Several cells
may be allocated on the same physical scaffold node. Their labels and pointers
are read through Circuit's local snapshots until the node is emitted.
"""
from scaffold_circuit import (Circuit, Value, Ref, EMPTY, NEW, PREVIOUS,
                              TRUE, FALSE, choose, conjunction, disjunction, neg)


class StackPool:
  def __init__(self, circuit, layout, payload=(None,)):
    self.circuit = circuit
    self.tags = tuple((name, index) for name, count in layout.items() for index in range(count))
    if not self.tags: raise ValueError("nonempty finite cell layout required")
    self.layout, self.used, self.payload = dict(layout), {}, tuple(payload)
    self.tag_ids = {tag: index for index, tag in enumerate(self.tags)}
    self.prefix = f"pool{len(getattr(circuit, '_pools', []))}"
    if not hasattr(circuit, '_pools'): circuit._pools = []
    circuit._pools.append(self)

  def key(self, tag, field):
    return f"{self.prefix}.c{self.tag_ids[tag]}.{field}"

  def scalar(self, ref, tag, field, domain, initial):
    # Tag equality guards are pairwise disjoint. Encode a field read as one
    # sum of products per bit instead of a long nest of binary multiplexers.
    # This also shares the validity test across all slots of this pool.
    default = Value.constant(initial).recode(domain)
    choices = []
    for variant in self.tags:
      condition = tag.eq(variant)
      if condition != FALSE:
        value = self.circuit.get(ref, self.key(variant, field), domain, initial)
        choices.append((condition, value))
    missing = neg(disjunction(*(condition for condition, _ in choices)))
    bits = tuple(disjunction(conjunction(missing, bit),
                   *(conjunction(condition, value.bits[index]) for condition, value in choices))
                 for index, bit in enumerate(default.bits))
    valid = disjunction(missing,
              *(conjunction(condition, value.valid) for condition, value in choices))
    return Value.encoded(domain, bits, valid)

  def pointer(self, ref, tag, field):
    answer = EMPTY
    for variant in self.tags:
      condition = tag.eq(variant)
      if condition != FALSE:
        value = self.circuit.get_ref(ref, self.key(variant, field))
        answer = Ref.select(condition, value, answer)
    return answer

  def allocate(self, name, previous, previous_tag, value, data, enabled, slot=None):
    index = self.used.get(name, 0) if slot is None else slot
    if index >= self.layout[name]:
      raise ValueError(f"finite stack cell layout exhausted for {name}")
    self.used[name] = max(self.used.get(name, 0), index + 1)
    tag = name, index
    self.circuit.put_ref(self.key(tag, "below"), Ref.select(enabled, previous,
      self.circuit.current_refs.get(self.key(tag, "below"), EMPTY)))
    self.circuit.put(self.key(tag, "tag"),
      Value.select(enabled, previous_tag, self.circuit.current_values.get(self.key(tag, "tag"),
        Value.constant(self.tags[0]))), self.tags, self.tags[0])
    self.circuit.put_ref(self.key(tag, "value"), Ref.select(enabled, value,
      self.circuit.current_refs.get(self.key(tag, "value"), EMPTY)))
    self.circuit.put(self.key(tag, "data"),
      Value.select(enabled, data, self.circuit.current_values.get(self.key(tag, "data"),
        Value.constant(self.payload[0]))), self.payload, self.payload[0])
    return NEW, Value.constant(tag).recode(self.tags)


class Stack:
  def __init__(self, pool, name, allocation_name=None):
    self.pool, self.name, self.circuit = pool, name, pool.circuit
    self.allocation_name = name if allocation_name is None else allocation_name
    self.root_key = f"{pool.prefix}.{name}.root"
    self.tag_key = f"{pool.prefix}.{name}.tag"
    self.top = self.circuit.get_ref(PREVIOUS, self.root_key)
    self.tag = self.circuit.get(PREVIOUS, self.tag_key, pool.tags, pool.tags[0])

  def empty(self):
    return neg(self.top.present())

  def peek(self):
    return (self.pool.pointer(self.top, self.tag, "value"),
            self.pool.scalar(self.top, self.tag, "data", self.pool.payload, self.pool.payload[0]))

  def pop(self, enabled=TRUE):
    value = self.peek()
    self.drop(enabled)
    return value

  def drop(self, enabled=TRUE):
    below = self.pool.pointer(self.top, self.tag, "below")
    tag = self.pool.scalar(self.top, self.tag, "tag", self.pool.tags, self.pool.tags[0])
    self.top = Ref.select(enabled, below, self.top)
    self.tag = Value.select(enabled, tag, self.tag)

  def push(self, value=EMPTY, data=None, enabled=TRUE, slot=None):
    if enabled == FALSE: return
    if data is None: data = Value.constant(self.pool.payload[0])
    top, tag = self.pool.allocate(self.allocation_name, self.top, self.tag, value, data, enabled, slot)
    self.top = Ref.select(enabled, top, self.top)
    self.tag = Value.select(enabled, tag, self.tag)

  def clear(self, enabled=TRUE):
    self.top = Ref.select(enabled, EMPTY, self.top)

  def copy_from(self, other, enabled=TRUE):
    if self.pool is not other.pool:
      raise ValueError("aliased stacks must share a declared finite cell pool")
    self.top = Ref.select(enabled, other.top, self.top)
    self.tag = Value.select(enabled, other.tag, self.tag)

  def finalize(self):
    self.circuit.put_ref(self.root_key, self.top)
    self.circuit.put(self.tag_key, self.tag)


class Tape:
  def __init__(self, circuit, name, alphabet, slots=1, blank="_", pool=None):
    self.circuit, self.name, self.alphabet, self.blank = circuit, name, tuple(alphabet), blank
    # Moving copies the focus symbol, never a stack cell. Left and right cell
    # tags therefore have separate finite domains, avoiding an unnecessary
    # cross-product during computed-cell reads.
    sizes = (slots, slots) if isinstance(slots, int) else tuple(slots)
    if pool is None:
      self.left, self.right = [Stack(StackPool(circuit, {name + side: count}, self.alphabet), name + side)
                               for side, count in zip((".l", ".r"), sizes)]
    else:
      # A Program may assign the same construction slot to mutually exclusive
      # tape moves. Stack roots stay distinct; only their cell storage is shared.
      self.left, self.right = [Stack(pool, name + side, "cells") for side in (".l", ".r")]
    self.symbol_key = name + ".symbol"
    self.focus = circuit.get(PREVIOUS, self.symbol_key, self.alphabet, blank)

  def write(self, value, enabled=TRUE):
    self.focus = Value.select(enabled, value.recode(self.alphabet), self.focus)

  def reset(self, enabled=TRUE):
    self.left.clear(enabled); self.right.clear(enabled)
    self.write(Value.constant(self.blank), enabled)

  def move(self, direction, enabled=TRUE, slot=None):
    if enabled == FALSE: return
    pushed, popped = (self.left, self.right) if direction == 1 else (self.right, self.left)
    if direction == -1: self.circuit.require(neg(popped.empty()), enabled)
    empty = popped.empty()
    _, value = popped.pop(enabled)
    if value.domain != self.alphabet:
      value = Value({char: value.eq(char) for char in self.alphabet})
      self.circuit.require(value.valid, conjunction(enabled, neg(empty)))
    pushed.push(data=self.focus, enabled=enabled, slot=slot)
    self.write(Value.select(empty, Value.constant(self.blank), value), enabled)

  def finalize(self):
    self.left.finalize(); self.right.finalize()
    self.circuit.put(self.symbol_key, self.focus)


class Counter:
  def __init__(self, pool, name, allocation_name=None):
    pools = pool if isinstance(pool, dict) else {"pos": pool, "neg": pool}
    self.pos, self.neg = (Stack(pools[side], name + "." + side, allocation_name)
                          for side in ("pos", "neg"))

  def positive(self):
    return neg(self.pos.empty())

  def negative(self):
    return neg(self.neg.empty())

  def zero(self):
    return conjunction(self.pos.empty(), self.neg.empty())

  def inc(self, enabled=TRUE, slot=None):
    empty = self.neg.empty()
    self.neg.drop(conjunction(enabled, neg(empty)))
    self.pos.push(enabled=conjunction(enabled, empty), slot=slot)

  def dec(self, enabled=TRUE, slot=None):
    empty = self.pos.empty()
    self.pos.drop(conjunction(enabled, neg(empty)))
    self.neg.push(enabled=conjunction(enabled, empty), slot=slot)

  def reset(self, enabled=TRUE):
    self.pos.clear(enabled); self.neg.clear(enabled)

  def copy_from(self, other, enabled=TRUE):
    self.pos.copy_from(other.pos, enabled); self.neg.copy_from(other.neg, enabled)

  def finalize(self):
    self.pos.finalize(); self.neg.finalize()


class Queue:
  NAMES = ("F", "B", "Fr", "Br", "WF", "WB", "B2")

  def __init__(self, pool, counter_pool, name, shared_slots=False):
    pools = pool if isinstance(pool, dict) else {role: pool for role in self.NAMES}
    counter_pools = counter_pool if isinstance(counter_pool, dict) else {"m": counter_pool, "c": counter_pool}
    self.circuit, self.name = pools["F"].circuit, name
    self.shared_slots = shared_slots
    allocation = "cells" if shared_slots else None
    self.s = {role: Stack(pools[role], f"{name}.{role}", allocation) for role in self.NAMES}
    self.m, self.c = (Counter(counter_pools[role], name + "." + role, allocation)
                      for role in ("m", "c"))
    self.phase_key = name + ".phase"
    self.phase = self.circuit.get(PREVIOUS, self.phase_key, ("idle", "rev", "copy"), "idle")

  def _phase(self, value, enabled):
    self.phase = Value.select(enabled, Value.constant(value), self.phase)

  def _slot(self, index=0):
    return index if self.shared_slots else None

  def push(self, value, enabled=TRUE):
    idle = self.phase.eq("idle")
    self.s["B"].push(value, enabled=conjunction(enabled, idle), slot=self._slot())
    self.s["B2"].push(value, enabled=conjunction(enabled, neg(idle)), slot=self._slot())
    self.c.dec(enabled, slot=self._slot())

  def pop(self, enabled=TRUE):
    self.circuit.require(neg(self.s["F"].empty()), enabled)
    value, _ = self.s["F"].pop(enabled)
    self.c.dec(enabled, slot=self._slot())
    rotating = conjunction(enabled, neg(self.phase.eq("idle")))
    self.m.dec(rotating, slot=self._slot())
    self._maybe_finish(rotating)
    return value

  def empty(self):
    return conjunction(self.s["F"].empty(), self.s["B"].empty(), self.phase.eq("idle"))

  def clear(self, enabled=TRUE):
    for stack in self.s.values(): stack.clear(enabled)
    self.m.reset(enabled); self.c.reset(enabled)
    self._phase("idle", enabled)

  def _start(self, enabled):
    self.s["WF"].copy_from(self.s["F"], enabled)
    self.s["WB"].copy_from(self.s["B"], enabled)
    for name in ("Fr", "Br", "B2"): self.s[name].clear(enabled)
    self.m.reset(enabled)
    self._phase("rev", enabled)

  def _maybe_finish(self, enabled):
    finish = conjunction(enabled, self.phase.eq("copy"), self.m.zero())
    self.s["F"].copy_from(self.s["Br"], finish)
    self.s["B"].copy_from(self.s["B2"], finish)
    for name in ("Br", "B2", "Fr"): self.s[name].clear(finish)
    self._phase("idle", finish)

  def _unit(self, enabled):
    reverse = conjunction(enabled, self.phase.eq("rev"))
    copy = conjunction(enabled, self.phase.eq("copy"))
    rear = conjunction(reverse, neg(self.s["WB"].empty()))
    front = conjunction(reverse, neg(self.s["WF"].empty()))
    value, _ = self.s["WB"].pop(rear)
    self.s["Br"].push(value, enabled=rear, slot=self._slot())
    self.c.inc(rear, slot=self._slot()); self.c.inc(rear, slot=self._slot(1))
    value, _ = self.s["WF"].pop(front)
    self.s["Fr"].push(value, enabled=front, slot=self._slot())
    self.m.inc(front, slot=self._slot())
    finished_reversal = conjunction(reverse, neg(disjunction(rear, front)))
    self._phase("copy", finished_reversal)
    self._maybe_finish(finished_reversal)
    copying = conjunction(copy, self.m.positive(), neg(self.s["Fr"].empty()))
    value, _ = self.s["Fr"].pop(copying)
    self.s["Br"].push(value, enabled=copying, slot=self._slot())
    self.m.dec(copying, slot=self._slot())
    self._maybe_finish(copy)

  def work(self, enabled=TRUE):
    if self.shared_slots:
      raise ValueError("shared queue cells require one work_unit per physical transition")
    for _ in range(3):
      self.work_unit(enabled)

  def work_unit(self, enabled=TRUE):
    """One rotation instruction; shared-slot users must schedule it explicitly."""
    self._start(conjunction(enabled, self.phase.eq("idle"), self.c.negative()))
    self._unit(conjunction(enabled, neg(self.phase.eq("idle"))))

  def copy_from(self, other, enabled=TRUE):
    for name in self.NAMES: self.s[name].copy_from(other.s[name], enabled)
    self.m.copy_from(other.m, enabled); self.c.copy_from(other.c, enabled)
    self.phase = Value.select(enabled, other.phase, self.phase)

  def finalize(self):
    for stack in self.s.values(): stack.finalize()
    self.m.finalize(); self.c.finalize()
    self.circuit.put(self.phase_key, self.phase)

File: scala/pal/src/main/scala/pal/ScaffoldCircuitStructs.scala
package pal

import scala.collection.immutable.VectorMap
import scala.collection.mutable

import Expr.{TRUE, FALSE}
import ScaffoldCircuit.{conjunction, disjunction, neg}
import Ref.{EMPTY, NEW, PREVIOUS}

/** Persistent stack and tape operations emitted as finite scaffold equations.
  *
  * All loops here range over a fixed construction-time cell layout. Several cells
  * may be allocated on the same physical scaffold node. Their labels and pointers
  * are read through Circuit's local snapshots until the node is emitted.
  *
  * Port of `scaffold_circuit_structs.py`. Python's `finalize()` methods are
  * `commit()` here (`Object.finalize` is deprecated on the JVM).
  */

/** A cell tag: the layout name and the slot index within that name. */
type CellTag = (String, Int)

/** A pool of finitely many cells, each with `below`, `tag`, `value` and `data` fields. */
final class StackPool(val circuit: Circuit, layoutFields: Iterable[(String, Int)], payloadValues: Seq[Any] = Vector(None)) {
  val layout: VectorMap[String, Int] = VectorMap.from(layoutFields)
  val tags: Vector[CellTag] = layout.toVector.flatMap { case (name, count) => (0 until count).map(index => (name, index)) }
  if (tags.isEmpty) { throw new IllegalArgumentException("nonempty finite cell layout required") }
  val used: mutable.LinkedHashMap[String, Int] = mutable.LinkedHashMap.empty
  val payload: Vector[Any] = payloadValues.toVector
  val tagIds: Map[CellTag, Int] = tags.zipWithIndex.toMap
  val prefix: String = s"pool${circuit.pools.size}"
  circuit.pools += this

  def key(tag: CellTag, field: String): String = s"$prefix.c${tagIds(tag)}.$field"

  /** Read scalar `field` of the cell selected by `tag`, defaulting to `initial` for no cell.
    *
    * Tag equality guards are pairwise disjoint. Encode a field read as one
    * sum of products per bit instead of a long nest of binary multiplexers.
    * This also shares the validity test across all slots of this pool.
    */
  def scalar[A](ref: Ref, tag: Value[?], field: String, domain: Seq[A], initial: A): Value[A] = {
    val default = Value.constant(initial).recode(domain)
    val choices = tags.flatMap { variant =>
      val condition = tag.eqTo(variant)
      if (condition != FALSE) { Some(condition -> circuit.get(ref, key(variant, field), domain, initial)) } else { None }
    }
    val missing = neg(disjunction(choices.map(_._1)*))
    val bits = default.bits.zipWithIndex.map { case (bit, index) =>
      disjunction((conjunction(missing, bit) +: choices.map { case (condition, value) => conjunction(condition, value.bits(index)) })*)
    }
    val valid = disjunction((missing +: choices.map { case (condition, value) => conjunction(condition, value.valid) })*)
    Value.encoded(domain, bits, valid)
  }

  /** Read pointer `field` of the cell selected by `tag`. */
  def pointer(ref: Ref, tag: Value[?], field: String): Ref = {
    var answer = EMPTY
    for (variant <- tags) {
      val condition = tag.eqTo(variant)
      if (condition != FALSE) {
        val value = circuit.getRef(ref, key(variant, field))
        answer = Ref.select(condition, value, answer)
      }
    }
    answer
  }

  /** Allocate the next (or the given) slot of `name`; returns the new cell and its tag. */
  def allocate(name: String, previous: Ref, previousTag: Value[?], value: Ref, data: Value[?], enabled: Expr, slot: Option[Int] = None): (Ref, Value[CellTag]) = {
    val index = slot.getOrElse(used.getOrElse(name, 0))
    if (index >= layout(name)) { throw new IllegalArgumentException(s"finite stack cell layout exhausted for $name") }
    used(name) = math.max(used.getOrElse(name, 0), index + 1)
    val tag: CellTag = (name, index)
    circuit.putRef(key(tag, "below"), Ref.select(enabled, previous, circuit.currentRefs.getOrElse(key(tag, "below"), EMPTY)))
    circuit.put(key(tag, "tag"),
      Value.select[Any](enabled, previousTag, circuit.currentValues.getOrElse(key(tag, "tag"), Value.constant(tags.head))),
      tags, tags.head)
    circuit.putRef(key(tag, "value"), Ref.select(enabled, value, circuit.currentRefs.getOrElse(key(tag, "value"), EMPTY)))
    circuit.put(key(tag, "data"),
      Value.select[Any](enabled, data, circuit.currentValues.getOrElse(key(tag, "data"), Value.constant(payload.head))),
      payload, payload.head)
    (NEW, Value.constant(tag).recode(tags))
  }
}

/** A persistent stack rooted in a pool cell; `commit()` writes the root back. */
final class Stack(val pool: StackPool, val name: String, allocation: Option[String] = None) {
  val circuit: Circuit = pool.circuit
  val allocationName: String = allocation.getOrElse(name)
  val rootKey: String = s"${pool.prefix}.$name.root"
  val tagKey: String = s"${pool.prefix}.$name.tag"
  var top: Ref = circuit.getRef(PREVIOUS, rootKey)
  var tag: Value[CellTag] = circuit.get(PREVIOUS, tagKey, pool.tags, pool.tags.head)

  def empty(): Expr = neg(top.present())

  def peek(): (Ref, Value[Any]) = {
    (pool.pointer(top, tag, "value"), pool.scalar(top, tag, "data", pool.payload, pool.payload.head))
  }

  def pop(enabled: Expr = TRUE): (Ref, Value[Any]) = {
    val value = peek()
    drop(enabled)
    value
  }

  def drop(enabled: Expr = TRUE): Unit = {
    val below = pool.pointer(top, tag, "below")
    val belowTag = pool.scalar(top, tag, "tag", pool.tags, pool.tags.head)
    top = Ref.select(enabled, below, top)
    tag = Value.select(enabled, belowTag, tag)
  }

  def push(value: Ref = EMPTY, data: Option[Value[Any]] = None, enabled: Expr = TRUE, slot: Option[Int] = None): Unit = {
    if (enabled != FALSE) {
      val payload = data.getOrElse(Value.constant(pool.payload.head))
      val (newTop, newTag) = pool.allocate(allocationName, top, tag, value, payload, enabled, slot)
      top = Ref.select(enabled, newTop, top)
      tag = Value.select(enabled, newTag, tag)
    }
  }

  def clear(enabled: Expr = TRUE): Unit = {
    top = Ref.select(enabled, EMPTY, top)
  }

  def copyFrom(other: Stack, enabled: Expr = TRUE): Unit = {
    if (pool ne other.pool) { throw new IllegalArgumentException("aliased stacks must share a declared finite cell pool") }
    top = Ref.select(enabled, other.top, top)
    tag = Value.select(enabled, other.tag, tag)
  }

  /** Python `finalize()`: write the root and tag back into the circuit. */
  def commit(): Unit = {
    circuit.putRef(rootKey, top)
    circuit.put(tagKey, tag)
  }
}

/** A two-stack tape with a focus symbol.
  *
  * Python accepted `slots` as one count for both sides or a `(left, right)`
  * pair; here the constructor takes the single count and `Tape.withSides`
  * takes the pair. Moving copies the focus symbol, never a stack cell. Left
  * and right cell tags therefore have separate finite domains, avoiding an
  * unnecessary cross-product during computed-cell reads.
  */
final class Tape private (
    val circuit: Circuit,
    val name: String,
    alphabetValues: Iterable[Any],
    sizes: (Int, Int),
    val blank: Any,
    pool: Option[StackPool]
) {

  /** `slots` cells on each side (Python `Tape(circuit, name, alphabet, slots=int, ...)`). */
  def this(circuit: Circuit, name: String, alphabet: Iterable[Any], slots: Int = 1, blank: Any = '_', pool: Option[StackPool] = None) = {
    this(circuit, name, alphabet, (slots, slots), blank, pool)
  }

  val alphabet: Vector[Any] = alphabetValues.toVector

  val (left: Stack, right: Stack) = {
    pool match {
      case None =>
        def side(suffix: String, count: Int): Stack = new Stack(new StackPool(circuit, Vector((name + suffix) -> count), alphabet), name + suffix)
        (side(".l", sizes._1), side(".r", sizes._2))
      case Some(shared) =>
        // A Program may assign the same construction slot to mutually exclusive
        // tape moves. Stack roots stay distinct; only their cell storage is shared.
        (new Stack(shared, name + ".l", Some("cells")), new Stack(shared, name + ".r", Some("cells")))
    }
  }

  val symbolKey: String = name + ".symbol"
  var focus: Value[Any] = circuit.get(PREVIOUS, symbolKey, alphabet, blank)

  def write(value: Value[Any], enabled: Expr = TRUE): Unit = {
    focus = Value.select(enabled, value.recode(alphabet), focus)
  }

  def reset(enabled: Expr = TRUE): Unit = {
    left.clear(enabled)
    right.clear(enabled)
    write(Value.constant(blank), enabled)
  }

  def move(direction: Int, enabled: Expr = TRUE, slot: Option[Int] = None): Unit = {
    if (enabled != FALSE) {
      val (pushed, popped) = if (direction == 1) { (left, right) } else { (right, left) }
      if (direction == -1) { circuit.require(neg(popped.empty()), enabled) }
      val empty = popped.empty()
      val (_, popValue) = popped.pop(enabled)
      val value = if (popValue.domain != alphabet) {
        val recoded = Value(alphabet.map(char => char -> popValue.eqTo(char)))
        circuit.require(recoded.valid, conjunction(enabled, neg(empty)))
        recoded
      } else {
        popValue
      }
      pushed.push(data = Some(focus), enabled = enabled, slot = slot)
      write(Value.select(empty, Value.constant(blank), value), enabled)
    }
  }

  /** Python `finalize()`. */
  def commit(): Unit = {
    left.commit()
    right.commit()
    circuit.put(symbolKey, focus)
  }
}

object Tape {

  /** Python `Tape(..., slots=(left, right), ...)`: different cell counts per side. */
  def withSides(circuit: Circuit, name: String, alphabet: Iterable[Any], slots: (Int, Int), blank: Any = '_', pool: Option[StackPool] = None): Tape = {
    new Tape(circuit, name, alphabet, slots, blank, pool)
  }
}

/** A signed unary counter: two stacks of which at most one is nonempty.
  *
  * The stacks are Python's `pos`/`neg` attributes, named `positiveStack` /
  * `negativeStack` here so that `neg` keeps meaning Boolean negation. They are
  * reassignable because `scaffold_window_stream` copies counters field-wise.
  */
final class Counter(pools: Map[String, StackPool], val name: String, allocationName: Option[String] = None) {

  def this(pool: StackPool, name: String, allocationName: Option[String]) = this(Map("pos" -> pool, "neg" -> pool), name, allocationName)

  def this(pool: StackPool, name: String) = this(pool, name, None)

  var positiveStack: Stack = new Stack(pools("pos"), name + ".pos", allocationName)
  var negativeStack: Stack = new Stack(pools("neg"), name + ".neg", allocationName)

  /** Python-name alias of `positiveStack`. */
  def pos: Stack = positiveStack

  def positive(): Expr = neg(positiveStack.empty())

  def negative(): Expr = neg(negativeStack.empty())

  def zero(): Expr = conjunction(positiveStack.empty(), negativeStack.empty())

  def inc(enabled: Expr = TRUE, slot: Option[Int] = None): Unit = {
    val empty = negativeStack.empty()
    negativeStack.drop(conjunction(enabled, neg(empty)))
    positiveStack.push(enabled = conjunction(enabled, empty), slot = slot)
  }

  def dec(enabled: Expr = TRUE, slot: Option[Int] = None): Unit = {
    val empty = positiveStack.empty()
    positiveStack.drop(conjunction(enabled, neg(empty)))
    negativeStack.push(enabled = conjunction(enabled, empty), slot = slot)
  }

  def reset(enabled: Expr = TRUE): Unit = {
    positiveStack.clear(enabled)
    negativeStack.clear(enabled)
  }

  def copyFrom(other: Counter, enabled: Expr = TRUE): Unit = {
    positiveStack.copyFrom(other.positiveStack, enabled)
    negativeStack.copyFrom(other.negativeStack, enabled)
  }

  /** Python `finalize()`. */
  def commit(): Unit = {
    positiveStack.commit()
    negativeStack.commit()
  }
}

object Counter {
  /** Python `Counter(pool, ...)` with a per-side pool dictionary. */
  def withPools(pools: Map[String, StackPool], name: String, allocationName: Option[String] = None): Counter = {
    new Counter(pools, name, allocationName)
  }
}

/** A FIFO queue built from seven stacks and two counters with amortized rotation. */
final class Queue(pools: Map[String, StackPool], counterPools: Map[String, StackPool], val name: String, val sharedSlots: Boolean = false) {

  def this(pool: StackPool, counterPool: StackPool, name: String, sharedSlots: Boolean) = {
    this(Queue.NAMES.map(role => role -> pool).toMap, Map("m" -> counterPool, "c" -> counterPool), name, sharedSlots)
  }

  def this(pool: StackPool, counterPool: StackPool, name: String) = this(pool, counterPool, name, false)

  val circuit: Circuit = pools("F").circuit
  private val allocation: Option[String] = if (sharedSlots) { Some("cells") } else { None }

  /** The seven role stacks (Python `self.s`). */
  val stacks: VectorMap[String, Stack] = VectorMap.from(Queue.NAMES.map(role => role -> new Stack(pools(role), s"$name.$role", allocation)))
  val m: Counter = new Counter(counterPools("m"), name + ".m", allocation)
  val c: Counter = new Counter(counterPools("c"), name + ".c", allocation)
  val phaseKey: String = name + ".phase"
  var phase: Value[String] = circuit.get(PREVIOUS, phaseKey, Vector("idle", "rev", "copy"), "idle")

  private def setPhase(value: String, enabled: Expr): Unit = {
    phase = Value.select(enabled, Value.constant(value), phase)
  }

  private def slotFor(index: Int = 0): Option[Int] = if (sharedSlots) { Some(index) } else { None }

  def push(value: Ref, enabled: Expr = TRUE): Unit = {
    val idle = phase.eqTo("idle")
    stacks("B").push(value, enabled = conjunction(enabled, idle), slot = slotFor())
    stacks("B2").push(value, enabled = conjunction(enabled, neg(idle)), slot = slotFor())
    c.dec(enabled, slot = slotFor())
  }

  def pop(enabled: Expr = TRUE): Ref = {
    circuit.require(neg(stacks("F").empty()), enabled)
    val (value, _) = stacks("F").pop(enabled)
    c.dec(enabled, slot = slotFor())
    val rotating = conjunction(enabled, neg(phase.eqTo("idle")))
    m.dec(rotating, slot = slotFor())
    maybeFinish(rotating)
    value
  }

  def empty(): Expr = conjunction(stacks("F").empty(), stacks("B").empty(), phase.eqTo("idle"))

  def clear(enabled: Expr = TRUE): Unit = {
    for (stack <- stacks.values) { stack.clear(enabled) }
    m.reset(enabled)
    c.reset(enabled)
    setPhase("idle", enabled)
  }

  private def start(enabled: Expr): Unit = {
    stacks("WF").copyFrom(stacks("F"), enabled)
    stacks("WB").copyFrom(stacks("B"), enabled)
    for (role <- Seq("Fr", "Br", "B2")) { stacks(role).clear(enabled) }
    m.reset(enabled)
    setPhase("rev", enabled)
  }

  private def maybeFinish(enabled: Expr): Unit = {
    val finish = conjunction(enabled, phase.eqTo("copy"), m.zero())
    stacks("F").copyFrom(stacks("Br"), finish)
    stacks("B").copyFrom(stacks("B2"), finish)
    for (role <- Seq("Br", "B2", "Fr")) { stacks(role).clear(finish) }
    setPhase("idle", finish)
  }

  private def unit(enabled: Expr): Unit = {
    val reverse = conjunction(enabled, phase.eqTo("rev"))
    val copy = conjunction(enabled, phase.eqTo("copy"))
    val rear = conjunction(reverse, neg(stacks("WB").empty()))
    val front = conjunction(reverse, neg(stacks("WF").empty()))
    val (rearValue, _) = stacks("WB").pop(rear)
    stacks("Br").push(rearValue, enabled = rear, slot = slotFor())
    c.inc(rear, slot = slotFor())
    c.inc(rear, slot = slotFor(1))
    val (frontValue, _) = stacks("WF").pop(front)
    stacks("Fr").push(frontValue, enabled = front, slot = slotFor())
    m.inc(front, slot = slotFor())
    val finishedReversal = conjunction(reverse, neg(disjunction(rear, front)))
    setPhase("copy", finishedReversal)
    maybeFinish(finishedReversal)
    val copying = conjunction(copy, m.positive(), neg(stacks("Fr").empty()))
    val (copyValue, _) = stacks("Fr").pop(copying)
    stacks("Br").push(copyValue, enabled = copying, slot = slotFor())
    m.dec(copying, slot = slotFor())
    maybeFinish(copy)
  }

  def work(enabled: Expr = TRUE): Unit = {
    if (sharedSlots) { throw new IllegalArgumentException("shared queue cells require one work_unit per physical transition") }
    for (_ <- 0 until 3) { workUnit(enabled) }
  }

  /** One rotation instruction; shared-slot users must schedule it explicitly. */
  def workUnit(enabled: Expr = TRUE): Unit = {
    start(conjunction(enabled, phase.eqTo("idle"), c.negative()))
    unit(conjunction(enabled, neg(phase.eqTo("idle"))))
  }

  def copyFrom(other: Queue, enabled: Expr = TRUE): Unit = {
    for (role <- Queue.NAMES) { stacks(role).copyFrom(other.stacks(role), enabled) }
    m.copyFrom(other.m, enabled)
    c.copyFrom(other.c, enabled)
    phase = Value.select(enabled, other.phase, phase)
  }

  /** Python `finalize()`. */
  def commit(): Unit = {
    for (stack <- stacks.values) { stack.commit() }
    m.commit()
    c.commit()
    circuit.put(phaseKey, phase)
  }
}

object Queue {
  val NAMES: Vector[String] = Vector("F", "B", "Fr", "Br", "WF", "WB", "B2")

  /** Python `Queue(pool_dict, counter_pool_dict, ...)`. */
  def withPools(pools: Map[String, StackPool], counterPools: Map[String, StackPool], name: String, sharedSlots: Boolean = false): Queue = {
    new Queue(pools, counterPools, name, sharedSlots)
  }
}
